### PR TITLE
Patterns: Add editing of pattern categories to site editor

### PR DESCRIPTION
**Pull request by glendaviesnz**
## What?
Adds the ability to edit categories to the site editor pattern editor.

## Why?
Currently there is no way to edit the categories of a pattern in the site editor.

## How?
Adds a category editing row to the right panel of the pattern editor page.

## Testing Instructions

- In the site editor add a new pattern with some categories
- Once pattern is added select the pattern and go into editor
- Check that a category editing box appears in right settings panel and that it works as expected for adding and removing categories
- Edit a template part and make sure the category panel does not appear
- Edit a Template, make a change and save it if it has no revisions, then confirm the revisions still show in the sidebar

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/a8305b64-a0a2-4571-9ae8-9ac624c988ba

### CODE REVIEW

File: packages/edit-site/src/components/sidebar-edit-mode/template-panel/pattern-categories.js [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+	unescapeString( termB ).toLowerCase();
+
+const termNamesToIds = ( names, terms ) => {
+	return names.map(
+		( termName ) =>
+			terms.find( ( term ) => isSameTermName( term.name, termName ) ).id
+	);
+};
+
+export default function PatternCategories( { post } ) {
```
Comment (glendaviesnz) [inline packages/edit-site/src/components/sidebar-edit-mode/template-panel/pattern-categories.js:60]: this involved a bit of copying from the editor package because of the different way we are handling editing posts in the site editor, it would be nice to reduce the duplication at some point but I think this would be better handled post 6.4 - I think this small amount of copy paste is ok in the meantime.

File: packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js [hunk 2]
```diff
@@ -64,6 +70,7 @@ export default function TemplatePanel() {
 			>
 				<LastRevision />
 			</PanelRow>
+			{ postType === 'wp_block' && <PatternCategories post={ record } /> }
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js]: Should we be using the recently consolidated constants here and in other places instead of `'wp_block'`?

File: packages/edit-site/src/components/sidebar-edit-mode/template-panel/pattern-categories.js (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar-edit-mode/template-panel/pattern-categories.js:60]: Sounds good, perhaps we should create an issue to follow up on this so it doesn't slip through the cracks.

**Review by aaronrobertshaw — COMMENTED** [review]
It will be good to get this one into 6.4 as the re-categorization of patterns was pretty awkward having to leave the site editor to edit the pattern via the post type editor.

The PR tests well. The only thing that caught me out was not tokenizing the incomplete tokens in the category field as we do in the pattern creation modal. 

I'm not sure if we should be consistent with the pattern creation modal or the normal post-type category field. For the site editor, I'd probably lean towards the modal but it can be a follow-up.

The fit in the sidebar was a little clunky but I found that was due to the revisions rendering an empty row. I pushed a commit that cleans up that one. So given that, it might be better to get a fresh set of eyes to give a final review on this PR.

| Before | After |
|---|---|
| Screenshot 2023-09-20 at 3 36 36 pm | Screenshot 2023-09-20 at 4 04 00 pm |

File: packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js]: If desired, perhaps it could be included in a code quality follow-up as suggested for refactoring the `PatternCategories` component.

Comment (glendaviesnz) [reply packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js]: Have switched to the constant.

**Comment by glendaviesnz** [comment]
> The only thing that caught me out was not tokenizing the incomplete tokens in the category field as we do in the pattern creation modal.

I have switched it to tokenize on blur to match the modal.

File: packages/edit-site/src/components/sidebar-edit-mode/template-panel/pattern-categories.js
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-edit-mode/template-panel/pattern-categories.js:1]: For future reference, this file is largely copied/based on https://github.com/WordPress/gutenberg/blob/08b6f1c0a887eeebaf203bbf7a3f96cedd3b618c/packages/editor/src/components/post-taxonomies/flat-term-selector.js.

File: packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js [hunk 3]
```diff
@@ -58,12 +64,10 @@ export default function TemplatePanel() {
 			>
 				<TemplateAreas />
 			</SidebarCard>
-			<PanelRow
```
Comment (kevin940726) [inline packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js:61]: Is this change necessary?

**Review by kevin940726 — APPROVED** [review]
Code-wise LGTM 👍

File: packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js (hunk 3 above)
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js:61]: The empty row when there are no revisions creates a bunch of empty space in the sidebar when it shouldn't be there as I noted in my [review](https://github.com/WordPress/gutenberg/pull/54640#pullrequestreview-1634713475)

File: packages/edit-site/src/components/sidebar-edit-mode/template-panel/pattern-categories.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar-edit-mode/template-panel/pattern-categories.js:1]: Thanks for adding the link. The copy pasta was flagged [earlier](https://github.com/WordPress/gutenberg/pull/54640#discussion_r1330863056) with the possibility of a dedicated follow-up issue discussed.

File: packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js (hunk 3 above)
Comment (kevin940726) [reply packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js:61]: Oh yeah I just saw that now 😅 . 👍
